modules/apps/transfer/types/denom.go

```diff
@@ -0,0 +1,108 @@
+package types
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"sort"
+	"strings"
+
+	errorsmod "cosmossdk.io/errors"
+
+	cmtbytes "github.com/cometbft/cometbft/libs/bytes"
+)
+
+// Validate performs a basic validation of the Denom fields.
+func (d Denom) Validate() error {
+	// NOTE: base denom validation cannot be performed as each chain may define
+	// its own base denom validation
+	if strings.TrimSpace(d.Base) == "" {
+		return fmt.Errorf("base denomination cannot be blank")
+	}
+
+	if len(d.Trace) != 0 {
+		trace := strings.Join(d.Trace, "/")
+		identifiers := strings.Split(trace, "/")
+
+		if err := validateTraceIdentifiers(identifiers); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Hash returns the hex bytes of the SHA256 hash of the Denom fields using the following formula:
+//
+// hash = sha256(tracePath + "/" + baseDenom)
+func (d Denom) Hash() cmtbytes.HexBytes {
+	hash := sha256.Sum256([]byte(d.GetFullPath()))
+	return hash[:]
+}
+
+// IBCDenom a coin denomination for an ICS20 fungible token in the format
+// 'ibc/{hash(tracePath + baseDenom)}'. If the trace is empty, it will return the base denomination.
+func (d Denom) IBCDenom() string {
+	if d.IsNative() {
+		return d.Base
+	}
+
+	return fmt.Sprintf("%s/%s", DenomPrefix, d.Hash())
+}
+
+// GetFullPath returns the full denomination according to the ICS20 specification:
+// tracePath + "/" + baseDenom
+// If there exists no trace then the base denomination is returned.
+func (d Denom) GetFullPath() string {
+	if d.IsNative() {
+		return d.Base
+	}
+
+	path := d.Trace[0]
+	for i := 1; i <= len(d.Trace)-1; i++ {
+		path = fmt.Sprintf("%s/%s", path, d.Trace[i])
+	}
+
```

Comment:
needs tests #6416, if anyone has any other implementation suggestions, I'm all ears

Reply:
Oh yea, this is great. Will update